### ISSUE TITLE
Support pipelining in Redis API

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5364]
As explained in Redis's documentation in https://redis.io/topics/pipelining, the Redis protocol inherently supports pipelining: The client may send several requests on the same connection before waiting for responses, the server will read all of them, and start all of them, and will then return the responses in one write (in the same order as the requests, which requires buffering of the responses).

Looking at our code, if I understand correctly it seems that  `redis_server::connection::process()` has no problem with the client sending multiple requests - but it just parses one, waits for it to be completed and write the reply - and only then starts to parse and process the next one. We could reduce the latency and the amount of work that we need to do on writing the response (as one write instead of many small writes), if we add full support for pipelining - namely - start multiple requests read from the same socket before waiting for all of them to complete.

This issue is about performance, not correctness. As far as I can tell, the existing behavior is correct, just less efficient than it could have been.

@fastio

**Comment by fastio** [comment]
@nyh I want to try to work around this feature that support pipelining in Redis.

My simple solution is described as follows:

* Allocate one fixed size queue when the connection is created.
* Parse as many requests on the connection as possible util the queue is full.
* Process the requests in the queue as many as possible util the queue is empty.

Please give me some advice. Thanks.

**Comment by avikivity** [comment]
Processing should be done in parallel when possible. This will overlap the latency of contacting other nodes.

One problem is that it's not always possible to process in parallel, requests for the same key should be processed serially, otherwise a SET a x followed by a GET a might execute GET a before the SET.

To serialize, we can use the same locking that LWT uses (@gleb-cloudius where is it?) Note that acquiring the lock, and sending the response, must be in the original order, but everything in between can be parallelized.

**Comment by gleb-cloudius** [comment]
On Sun, Dec 15, 2019 at 05:47:38AM -0800, Avi Kivity wrote:
> Processing should be done in parallel when possible. This will overlap the latency of contacting other nodes.
> 
> One problem is that it's not always possible to process in parallel, requests for the same key should be processed serially, otherwise a SET a x followed by a GET a might execute GET a before the SET.
> 
> To serialize, we can use the same locking that LWT uses (@gleb-cloudius where is it?) Note that acquiring the lock, and sending the response, must be in the original order, but everything in between can be parallelized.
> 
It is in a paxos_state::_locks. Currently taken by
paxos_state::prepare() and paxos_state::accept() but I am moving it up
the call chain. FWIW it is per token, not per key.

--
			Gleb.

**Comment by avikivity** [comment]
That's okay for redis too

**Comment by fastio** [comment]
@avikivity Thanks for your advice.

**Comment by fastio** [comment]
How to guarantee the Redis requests for the same key should be processed serially?

The original Redis server is a single-process single-threaded model. The requests from the
client are processed in order as they recieved (FIFO). In this scenario, the requests for the
same key always processed serially.

Scylla runs single-threaded, sharded per hyperthread (Shared-nothing architecture).
Every shard executes the redis command for requests in parallel. So the requests for
the same key may  be processed by different shard in parallel. If it happens, the client
will get wrong result obviously.

How to make the requests for the same key  be processed serially?

**Approach 1**:  Global locks
In detail, the global lock should be abtained by arbitrary shard which prcessing the Redis requests.
The requests for the same key proccessed by different shards will queue on the lock (global). 
In the server scope, there is a locks of map (instance of std::unordered_map<string, semaphore>).
This approach is very simple. But it's constrary to the Share-Nothing philosophy, and detrimental to performance.

> If every shard holds its own locks, it's not avoid that the redis commands for the same key are
executed concurrent by different shards. In that case, a SET a x followed by a GET a might execute GET a before the SET.

**Approach 2**: Locks per shard.
In this approach, every shard holds its own locks ( the thread local instance of std::unordered_map<string, seastar::semaphore>). The request recieved by any shard will be submit to the target shard which id is equal to value of  std::hash(key) / smp::count. The presudo code is as follow:
* Parse the Redis request.
* Get the key from the request.
* Calculate the hash of the key.
* Submit the request to the target shard. 
``` return submit_to( std::hash(req.get_key()) / smp::count, [req = std::move(req)] { return process(req) }); ```

When the target shard recieving the redis request from other shard or itself, obtains the locks (per shard), and processes  the request. Obviously, the requests for the same key will be processed on
the only shard, and protected by locks. So there requests processed in order.
 
As described above, I want use the Approach 2. 
ping @nyh @avikivity

**Comment by dorlaor** [comment]
@fastio the lock per shard makes sense since each shard owns its keys and forward requests to the others. Better have @nyh @avikivity @penberg  here

**Comment by fastio** [comment]
@dorlaor  Yes, it works.

**Comment by nyh** [comment]
@fastio I think this discussion mixes up two separate issues...

The title of this issue is **pipelining**. In that case, the entire pipeline is already received on the same shard, and the locking can happen on this shard, no matter which node or shard each key "belongs" to. There is no need for any global locks or bouncing requests between shards.

A **different issue** is what happens when multiple unrelated requests are received from different clients for the same key concurrently. Especially INCR requests. This problem isn't just multiple shards (which as you said, you can solve by forwarding the request, or just the lock, or one specific shard - as we did in LWT with storage_proxy::cas_shard(token)). In the general case, you can also have concurrent modifications to the same key arriving on different nodes! I don't think you want to solve this problem yet, right? (by the way, LWT is one solution to this problem, but the current implementation comes with a hefty performance penalty).

**Comment by avikivity** [comment]
@nyh still we should not serialize requests within a pipeline if we can avoid it.

e.g. a pipeline that does

 INCR X
 INCR Y
 INCR X
 INCR Y

We should not run two INCR X in parallel, since this will produce wrong results and/or LWT contention, but we should run INCR x and INCR y in parallel.

So we can have a per-connection lock array:

   std::unordered_map<dht::token, semaphore> _locks;

each operation should serialize on a lock for the key's token, so two INCR X operations are executed in order.

Note: after a lock is released, we need to drop it from the map, otherwise the map will grow continuously. So perhaps we need something better than the map I gave as an example.

**Comment by nyh** [comment]
@avikivity of course. I didn't say we don't need to do this locking - we do. Just that it just needs to happen inside the shard (even the connection as you said) - and doesn't need to be cross-shard in any way like in the two options that @fastio and @dorlaor were discussing.

The map you described (with the easy optimization of dropping the semaphore when it returns to its ground state) sounds exactly right.

By the way, I would add another semaphore, limiting the **total** number of parallel request started by the connection - so we don't start even more requests before finishing some older ones. I don't know if this semaphore should be in the number of requests, or its memory use - like we do in "admission control" in CQL.

**Comment by avikivity** [comment]
Probably it should be global to the shard (and not to the connection), we we don't have (many connections) * (deep pipeline) == too many requests

**Comment by nyh** [comment]
> Probably it should be global to the shard (and not to the connection), we we don't have (many connections) * (deep pipeline) == too many requests

Makes sense. But my point was that there is no need to forward requests to specific shards, as @fastio suggested in his "Approach 2". This request-forwarding thing would have been great on a single-node installation (where it would have provided global serialization of requests to the same key), but would be of no help at all on multi-node installations.

**Comment by nyh** [comment]
In commit 09dc285b4a47cdbcb83025e52ba7090593f825f4, the Redis API support was removed from Scylla, so closing the issues related to it.
